src-c/native/whirlpool.c

```diff
@@ -0,0 +1,729 @@
+		0xbd5d10f4cb3e0567ULL,
+		0xe427418ba77d95d8ULL,
+		0xfbee7c66dd17479eULL,
+		0xca2dbf07ad5a8333ULL
+	};
+
+	/* map the message buffer to a block */
+	for (i = 0; i < 8; i++) {
+		/* store K^0 and xor it with the intermediate hash state */
+		K[0][i] = hash[i];
+		state[0][i] = be64_to_cpu(p_block[i]) ^ hash[i];
+		hash[i] = state[0][i];
+	}
+
+	/* iterate over algorithm rounds */
+	for (i = 0; i < number_of_rounds; i++)
+	{
+		/* compute K^i from K^{i-1} */
+		K[m ^ 1][0] = WHIRLPOOL_OP(K[m], 0) ^ rc[i];
+		K[m ^ 1][1] = WHIRLPOOL_OP(K[m], 1);
+		K[m ^ 1][2] = WHIRLPOOL_OP(K[m], 2);
+		K[m ^ 1][3] = WHIRLPOOL_OP(K[m], 3);
+		K[m ^ 1][4] = WHIRLPOOL_OP(K[m], 4);
+		K[m ^ 1][5] = WHIRLPOOL_OP(K[m], 5);
+		K[m ^ 1][6] = WHIRLPOOL_OP(K[m], 6);
+		K[m ^ 1][7] = WHIRLPOOL_OP(K[m], 7);
+
+		/* apply the i-th round transformation */
+		state[m ^ 1][0] = WHIRLPOOL_OP(state[m], 0) ^ K[m ^ 1][0];
+		state[m ^ 1][1] = WHIRLPOOL_OP(state[m], 1) ^ K[m ^ 1][1];
+		state[m ^ 1][2] = WHIRLPOOL_OP(state[m], 2) ^ K[m ^ 1][2];
+		state[m ^ 1][3] = WHIRLPOOL_OP(state[m], 3) ^ K[m ^ 1][3];
+		state[m ^ 1][4] = WHIRLPOOL_OP(state[m], 4) ^ K[m ^ 1][4];
+		state[m ^ 1][5] = WHIRLPOOL_OP(state[m], 5) ^ K[m ^ 1][5];
+		state[m ^ 1][6] = WHIRLPOOL_OP(state[m], 6) ^ K[m ^ 1][6];
+		state[m ^ 1][7] = WHIRLPOOL_OP(state[m], 7) ^ K[m ^ 1][7];
+
+		m = m ^ 1;
+	}
+
+	/* apply the Miyaguchi-Preneel compression function */
+	hash[0] ^= state[0][0];
+	hash[1] ^= state[0][1];
+	hash[2] ^= state[0][2];
+	hash[3] ^= state[0][3];
+	hash[4] ^= state[0][4];
+	hash[5] ^= state[0][5];
+	hash[6] ^= state[0][6];
+	hash[7] ^= state[0][7];
+}
+
+/**
+ * Calculate message hash.
+ * Can be called repeatedly with chunks of the message to be hashed.
+ *
+ * @param ctx the algorithm context containing current hashing state
+ * @param msg message chunk
+ * @param size length of the message chunk
+ */
+void digestif_whirlpool_update(struct whirlpool_ctx* ctx, uint8_t *data, uint32_t len)
+{
+	unsigned int index, to_fill;
+
+	/* check for partial buffer */
+	index = (unsigned int) (ctx->sz & 0x3f);
+	to_fill = 64 - index; // XXX
+
+	ctx->sz += len;
+
+	/* process partial buffer if there's enough data to make a block */
+	if (index && len >= to_fill) {
+		memcpy(ctx->buf + index, data, to_fill);
+		whirlpool_do_chunk(ctx->h, (uint64_t*)ctx->buf);
+		len -= to_fill;
+		data += to_fill;
+		index = 0;
+	}
+
+	/* process as much 128-block as possible */
+	for (; len >= 64; len -= 64, data += 64)
+		whirlpool_do_chunk(ctx->h, (uint64_t*)data);
+
+	/* append data into buf */
+	if (len)
+		memcpy(ctx->buf + index, data, len);
+}
+
+/**
+ * Store calculated hash into the given array.
+ *
+ * @param ctx the algorithm context containing current hashing state
+ * @param result calculated hash in binary form
+ */
+void digestif_whirlpool_finalize(struct whirlpool_ctx* ctx, uint8_t *out)
+{
+    uint32_t i, index;
+	uint64_t* msg64 = (uint64_t*)ctx->buf;
+	index = (unsigned int) (ctx->sz & 0x3f);
```

Comment:
doesn't really matter since `& 0x3f` is less than 16 bits, but shouldn't this be a `(uint32_t)` cast?

Reply:
Indeed this should be an uint32_t cast